File: src/main/java/com/github/cao/awa/conium/component/value/ConiumValueCreator.java
package com.github.cao.awa.conium.component.value;

import com.github.cao.awa.sinuatum.manipulate.Manipulate;
import com.google.gson.JsonElement;

public interface ConiumValueCreator<T> {
    T createValue(JsonElement element);

    default T castValue(Object value) {
        if (value instanceof JsonElement jsonElement) {
            return createValue(jsonElement);
        }
        return Manipulate.cast(value);
    }
}

File: common/src/main/kotlin/com/github/cao/awa/conium/component/value/ConiumValueCreator.kt
package com.github.cao.awa.conium.component.value

import com.github.cao.awa.conium.kotlin.extent.manipulate.doCast
import com.google.gson.JsonElement

fun interface ConiumValueCreator<T> {
    fun createValue(element: JsonElement): T

    fun <X> castValue(value: X): T {
        if (value is JsonElement) {
            return createValue(value)
        }
        return value.doCast()
    }
}
